Arrowgene.Ddon.GameServer/Characters/JobManager.cs

```diff
@@ -317,6 +405,115 @@ public void RemoveSkill(IDatabase database, CharacterCommon character, JobId job
+            },
+            [JobId.SpiritLancer] = new List<uint[]>() {
+                    new uint[] { LearnedNormalSkills.SpiritLancer.ThrustMore },
+                    new uint[] { LearnedNormalSkills.SpiritLancer.Aid },
+                    new uint[] { LearnedNormalSkills.SpiritLancer.FallMove }
+            },
+            [JobId.Warrior] = new List<uint[]>() {
+                    new uint[] { LearnedNormalSkills.Warrior.Devastate },
+                    new uint[] { LearnedNormalSkills.Warrior.SavageLash },
+                    new uint[] { LearnedNormalSkills.Warrior.ControlledFall }
+            },
+        };
+
+        public void UnlockLearnedNormalSkill(IDatabase database, GameClient client, CharacterCommon character, JobId Job, uint SkillIndex)
+        {
+            CDataCharacterJobData characterJobData = character.CharacterJobDataList.Where(cjd => cjd.Job == Job).Single();
+
+            if (!gLearnedNormalSkillMap.ContainsKey(Job) || SkillIndex == 0 || ((SkillIndex - 1) > gLearnedNormalSkillMap[Job].Count()))
+            {
+                // Something strange happened, either there is a new job (unlikely)
+                // or there is a missing skill, or someone tried to craft a custom
+                // packet to the server. For now, just return back the same
+                // data requested so the client doesn't hang.
+                var S2CSkillLearnNormalSkillRes = new S2CSkillLearnNormalSkillRes()
+                {
+                    Job = Job,
+                    SkillIndex = 0,
+                    NewJobPoint = characterJobData.JobPoint,
+                };
+
+                client.Send(S2CSkillLearnNormalSkillRes);
+                return;
+            }
+
+            List<uint[]> SkillMapping = gLearnedNormalSkillMap[Job];
+            foreach (uint SkillNo in SkillMapping[(int)(SkillIndex - 1)])
+            {
+                List<CDataNormalSkillParam> Matches = character.LearnedNormalSkills.Where(skill => skill != null && skill.Job == Job && skill.SkillNo == SkillNo).ToList();
+                if (Matches.Count() == 0)
+                {
+                    CDataNormalSkillParam NewSkill = new CDataNormalSkillParam()
+                    {
+                        Job = Job,
+                        Index = SkillIndex, // 1, 2, 3 based offset from packet
+                        SkillNo = SkillNo,  // Related to offsets in gNormalSkillMap
+                    };
+
+                    character.LearnedNormalSkills.Add(NewSkill);
+                    database.InsertIfNotExistsNormalSkillParam(character.CommonId, NewSkill);
+                }
+            }
+
+            // TODO: Subtract Job Points
+
+            var Result = new S2CSkillLearnNormalSkillRes()
+            {
+                Job = Job,
+                SkillIndex = SkillIndex,
+                NewJobPoint = characterJobData.JobPoint,
+            };
+
+            client.Send(Result);
```

Comment:
Have you tested if other players see the new skill when unlocking it while inside a party? There's a chance you gotta send an NTC packet informing of the new unlocked skill to the other party members

Reply:
I've been playing with a friend using this patch and it seems our skills are working. Is there supposed to be a visual effect when it happens or is it more like they will see me using the skill?

Reply:
In the PS4 build I noticed there's the packet with ID S2C_NORMAL_SKILL_LEARN_NTC. There's a chance this is meant to be sent, however if you don't notice any issues or broken animations maybe we can afford to skip sending it for now

Reply:
This `S2C_NORMAL_SKILL_LEARN_NTC` packet, it would be another packet I would send after sending the `S2CSkillLearnNormalSkillRes`?

Reply:
Yes, check equipping Custom Skills and Abilities in JobManager, right next to the line that sends the response, the NTC packet is sent to all the party members to inform them of the change.